# Segmenting and Clustering Neighborhoods in Toronto

This is the notebook for the peer graded assignment of the week 3 of the applied data science capstone course.

__Instructions__

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment.


2. Hints for scraping Notebook


3. To create the above dataframe:
The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.
If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.


4. Submit a link to your Notebook on your Github repository.(10 marks)

## Note: Folium maps are not rendered in GitHub, so please to see the full content displayed with interactive maps click [here](https://nbviewer.jupyter.org/github/saulovillasenor/ibm_data_science_professional_certificate/blob/main/course10_applied_data_science_capstone/week3_neighborhood_segmentation_and_clustering/neighborhoods_toronto.ipynb).

In [8]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim
from pandas.io.json import json_normalize
import folium 
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

## Scrapying content from wikipedia

In [28]:
link = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
response = requests.get(link)
soup = BeautifulSoup(response.text, "html.parser")

In [48]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                            'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                            'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                            'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [53]:
print(df.shape)
print(df.nunique())

(103, 3)
PostalCode      103
Borough          15
Neighborhood    103
dtype: int64


There are a total of 103 roles in the dataframe. This means that Toronto has 103 different postal codes for their 103 neighbourhoods and 15 different boroughs. 

## Obtaining the latitude and longitude of the neighbourhoods 

In [73]:
df_coordinates = pd.read_csv(r"\Users\saulo\Downloads\Geospatial_Coordinates.csv")
df_coordinates

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [76]:
df_join = df.join(df_coordinates.set_index(['Postal Code']), on=['PostalCode'])
df_join

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## Explore and cluster the neighborhoods in Toronto

In [80]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(f"The geograpical coordinates of Toronto are {latitude}, {longitude}")

The geograpical coordinates of Toronto are 43.6534817, -79.3839347


### Create a map of Toronto with neighborhoods superimposed on top.

In [90]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_join['Latitude'], df_join['Longitude'], df_join['Borough'], df_join['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Folium is a great visualization library. However, for illustration purposes, let's simplify the above map and segment and cluster only the boroughs which have the word "Toronto" in it. So let's slice the original dataframe and create a new dataframe of Toronto

In [93]:
toronto_data = df_join[df_join['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_data

PostalCode                 Borough  \
0         M5A        Downtown Toronto   
1         M5B        Downtown Toronto   
2         M5C        Downtown Toronto   
3         M4E            East Toronto   
4         M5E        Downtown Toronto   
5         M5G        Downtown Toronto   
6         M6G        Downtown Toronto   
7         M5H        Downtown Toronto   
8         M6H            West Toronto   
9         M4J  East York/East Toronto   
10        M5J        Downtown Toronto   
11        M6J            West Toronto   
12        M4K            East Toronto   
13        M5K        Downtown Toronto   
14        M6K            West Toronto   
15        M4L            East Toronto   
16        M5L        Downtown Toronto   
17        M4M            East Toronto   
18        M4N         Central Toronto   
19        M5N         Central Toronto   
20        M4P         Central Toronto   
21        M5P         Central Toronto   
22        M6P            West Toronto   
23        M4R         Central Toronto   
24        M5R         Central Toronto   
25        M6R            West Toronto   
26        M4S         Central Toronto   
27        M5S        Downtown Toronto   
28        M6S            West Toronto   
29        M4T         Central Toronto   
30        M5T        Downtown Toronto   
31        M4V         Central Toronto   
32        M5V        Downtown Toronto   
33        M4W        Downtown Toronto   
34        M5W  Downtown Toronto Stn A   
35        M4X        Downtown Toronto   
36        M5X        Downtown Toronto   
37        M4Y        Downtown Toronto   
38        M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Richmond, Adelaide, King  43.650571 -79.384568  
8                        Dufferin, Dovercourt Village  43.669005 -79.442259  
9                                  The Danforth  East  43.685347 -79.338106  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                    

Let's visualize the neighborhoods in these boroughs.

In [99]:
map2 = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = f'{neighborhood}, {borough}'
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map2)  
    
map2

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [100]:
CLIENT_ID = '' # Foursquare ID
CLIENT_SECRET = '' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

# My credentials were removed for security reasons

#### Let's explore the first neighborhood in our dataframe.

Getting the neighborhood's name.

In [101]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.

In [102]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print(f'Latitude and longitude values of {neighborhood_name} are {neighborhood_latitude}, {neighborhood_longitude}.')

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


__Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.__

First, let's create the GET request URL. Name your URL url.

In [1]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
#url  display URL

In [2]:
results = requests.get(url).json()

In [105]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [106]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [107]:
print(f'{nearby_venues.shape[0]} venues were returned by Foursquare.')

44 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Toronto

In [108]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [109]:

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

In [111]:
print(toronto_venues.shape)
toronto_venues.head()

(1601, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

In [113]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Central Bay Street                                                     69   
Christie                                                               16   
Church and Wellesley                                                   80   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           16   
Enclave of M4L                                                         16   
Enclave of M5E                                                         97   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              65   
Lawrence Park                                                           5   
Little Portugal, Trinity                                               42   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     22   
Parkdale, Roncesvalles                                                 14   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     32   
St. James Town                                                         87   
St. James Town, Cabbagetown                                            47   
Studio District                                                        36   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth  East                                                      4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            23   
CN Tower, King and Spadina, Railway Lands, Harb...                      14   
Central Bay Street                                                      69   
Christie                                                                16   
Church and Wellesley                                                    80   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                                                         9   
Dufferin, Dovercourt Village        

In [118]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


## 3. Analyse each neighbourhood

In [121]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot

Yoga Studio  Adult Boutique  Airport  Airport Food Court  \
0               0               0        0                   0   
1               0               0        0                   0   
2               0               0        0                   0   
3               0               0        0                   0   
4               0               0        0                   0   
...           ...             ...      ...                 ...   
1596            0               0        0                   0   
1597            0               0        0                   0   
1598            1               0        0                   0   
1599            0               0        0                   0   
1600            0               0        0                   0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1596               0                0                 0                    0   
1597               0                0                 0                    0   
1598               0                0                 0                    0   
1599               0                0                 0                    0   
1600               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Theater  Theme Restaurant  \
0                0         0  ...        0                 0   
1                0         0  ...        0                 0   
2                0         0  ...        0                 0   
3                0         0  ...        0                 0   
4                0         0  ...        0                 0   
...            ...       ...  ...      ...               ...   
1596             0         0  ...        0                 0   
1597             0         0  ...        0                 0   
1598             0         0  ...        0                 0   
1599             0         0  ...        0                 0   
1600             0         0  ...        0                 0   

      Tibetan Restaurant  Toy / Game Store  Trail  Train Station  \
0                      0                 0      0              0   
1                      0                 0      0              0   
2                      0                 0      0              0   
3                      0                 0      0              0   
4                      0                 0      0              0   
...                  ...               ...    ...            ...   
1596                   0                 0      0              0   
1597                   0                 0      0              0   
1598                   0                 0      0              0   
1599                   0                 0      0              0   
1600                   0                 0      0              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1596                              0                 0                      0   
1597                      

In [128]:
toronto_onehot.describe()

Yoga Studio  Adult Boutique      Airport  Airport Food Court  \
count  1601.000000     1601.000000  1601.000000         1601.000000   
mean      0.007495        0.000625     0.000625            0.000625   
std       0.086277        0.024992     0.024992            0.024992   
min       0.000000        0.000000     0.000000            0.000000   
25%       0.000000        0.000000     0.000000            0.000000   
50%       0.000000        0.000000     0.000000            0.000000   
75%       0.000000        0.000000     0.000000            0.000000   
max       1.000000        1.000000     1.000000            1.000000   

       Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
count     1601.000000      1601.000000       1601.000000          1601.000000   
mean         0.001249         0.000625          0.001249             0.011243   
std          0.035333         0.024992          0.035333             0.105468   
min          0.000000         0.000000          0.000000             0.000000   
25%          0.000000         0.000000          0.000000             0.000000   
50%          0.000000         0.000000          0.000000             0.000000   
75%          0.000000         0.000000          0.000000             0.000000   
max          1.000000         1.000000          1.000000             1.000000   

       Antique Shop     Aquarium  ...      Theater  Theme Restaurant  \
count   1601.000000  1601.000000  ...  1601.000000       1601.000000   
mean       0.001874     0.003123  ...     0.008120          0.000625   
std        0.043261     0.055814  ...     0.089772          0.024992   
min        0.000000     0.000000  ...     0.000000          0.000000   
25%        0.000000     0.000000  ...     0.000000          0.000000   
50%        0.000000     0.000000  ...     0.000000          0.000000   
75%        0.000000     0.000000  ...     0.000000          0.000000   
max        1.000000     1.000000  ...     1.000000          1.000000   

       Tibetan Restaurant  Toy / Game Store        Trail  Train Station  \
count         1601.000000       1601.000000  1601.000000    1601.000000   
mean             0.000625          0.000625     0.002498       0.001874   
std              0.024992          0.024992     0.049937       0.043261   
min              0.000000          0.000000     0.000000       0.000000   
25%              0.000000          0.000000     0.000000       0.000000   
50%              0.000000          0.000000     0.000000       0.000000   
75%              0.000000          0.000000     0.000000       0.000000   
max              1.000000          1.000000     1.000000       1.000000   

       Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
count                    1601.000000       1601.000000            1601.000000   
mean                        0.009994          0.001249               0.004372   
std                         0.099499          0.035333               0.065999   
min                         0.000000          0.000000               0.000000   
25%                         0.000000          0.000000               0.000000   
50%                         0.000000          0.000000               0.000000   
75%                         0.000000          0.000000               0.000000   
max                         1.000000          1.000000               1.000000   

          Wine Bar  
count  1601.000000  
mean      0.004372  
std       0.065999  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000  

[8 rows x 236 columns]

__Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category__

In [129]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
3                                  Central Bay Street     0.014493   
4                                            Christie     0.000000   
5                                Church and Wellesley     0.025000   
6                      Commerce Court, Victoria Hotel     0.000000   
7                                          Davisville     0.000000   
8                                    Davisville North     0.000000   
9                        Dufferin, Dovercourt Village     0.000000   
10                                     Enclave of M4L     0.062500   
11                                     Enclave of M5E     0.010309   
12             First Canadian Place, Underground city     0.000000   
13                           Forest Hill North & West     0.000000   
14                           Garden District, Ryerson     0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000   
16                      High Park, The Junction South     0.000000   
17                     India Bazaar, The Beaches West     0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000   
19                                      Lawrence Park     0.000000   
20                           Little Portugal, Trinity     0.023810   
21                        Moore Park, Summerhill East     0.000000   
22                                 North Toronto West     0.045455   
23                             Parkdale, Roncesvalles     0.000000   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.031250   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                    Studio District     0.027778   
32  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
33                The Annex, North Midtown, Yorkville     0.000000   
34                                        The Beaches     0.000000   
35                                 The Danforth  East     0.000000   
36                       The Danforth West, Riverdale     0.023810   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.030303   

    Adult Boutique   Airport  Airport Food Court  Airport Lounge  \
0           0.0000  0.000000            0.000000        0.000000   
1           0.0000  0.000000            0.000000        0.000000   
2           0.0000  0.071429            0.071429        0.142857   
3           0.0000  0.000000            0.000000        0.000000   
4           0.0000  0.000000            0.000000        0.000000   
5           0.0125  0.000000            0.000000        0.000000   
6           0.0000  0.000000            0.000000        0.000000   
7           0.0000  0.000000            0.000000        0.000000   
8           0.0000  0.000000            0.000000        0.000000   
9           0.0000  0.000000            0.000000        0.000000   
10          0.0000  0.000000            0.000000        0.000000   
11          0.0000  0.000000            0.000000        0.000000   
12          0.0000  0.000000            0.000000        0.000000   
13          0.0000  0.000000            0.000000        0.000000   
14          0.0000  0.000000            0.000000        0.000000   
15          0.0000  0.000000            0.000000        0.000000   
16          0.0000  0.000000            0.000000        0.000000   
17          0.00

__Let's print each neighborhood along with the top 5 most common venues__

In [132]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0  Cocktail Bar  0.05
1   Coffee Shop  0.05
2        Bakery  0.05
3   Cheese Shop  0.04
4    Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.09
2  Breakfast Spot  0.09
3       Pet Store  0.04
4   Burrito Place  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.14
1  Airport Terminal  0.14
2             Plane  0.07
3   Harbor / Marina  0.07
4     Boat or Ferry  0.07


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.16
1       Sandwich Place  0.06
2                 Café  0.06
3   Italian Restaurant  0.04
4  Japanese Restaurant  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3      Nightclub  0.06
4    Coffee Shop  0.06


----Church and

__Let's put that into a pandas dataframe__

First, let's write a function to sort the venues in descending order.

In [133]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [134]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                Bakery   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop          Cocktail Bar    Seafood Restaurant   
1        Breakfast Spot           Coffee Shop                   Gym   
2      Airport Terminal         Boat or Ferry      Sculpture Garden   
3        Sandwich Place                  Café    Italian Restaurant   
4                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar              Pharmacy            Restaurant   
1                Bakery               Stadium         Burrito Place   
2               Airport    Airport Food Court       Airport Service   
3           Salad Place         Deli / Bodega            Restaurant   
4           Coffee Shop            Restaurant    Italian Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market           Cheese Shop         Clothing Store  
1            Restaurant          Climbing Gym              Pet Store  
2       Harbor / Marina                   Bar            Coffee Shop  
3       Bubble Tea Shop   Japanese Restaurant           Burger Joint  
4    Athletics & Sports           Candy Store              Nightclub

## 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [135]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [136]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M4E      East Toronto                The Beaches  43.676357   
4        M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                  Park   
1 -79.378937               0           Coffee Shop        Clothing Store   
2 -79.375418               0           Coffee Shop                  Café   
3 -79.293031               2                   Pub                 Trail   
4 -79.373306               0                Bakery           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub                Bakery        Breakfast Spot   
1                 Hotel       Bubble Tea Shop                  Café   
2            Restaurant                Bakery        Cosmetics Shop   
3     Health Food Store              Wine Bar      Doner Restaurant   
4          Cocktail Bar    Seafood Restaurant              Beer Bar   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Theater                  Café     French Restaurant   
1        Cosmetics Shop   Japanese Restaurant        Lingerie Store   
2          Cocktail Bar        Clothing Store                   Gym   
3                 Diner        Discount Store   Distribution Center   
4              Pharmacy            Restaurant        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0        Chocolate Shop             Beer Store  
1  Fast Food Restaurant       Ramen Restaurant  
2                  Park    Moroccan Restaurant  
3               Dog Run    Dumpling Restaurant  
4           Cheese Shop         Clothing Store

Finally, let's visualize the resulting clusters

In [139]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

__Cluster 1__

In [140]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         Downtown Toronto               0           Coffee Shop   
1         Downtown Toronto               0           Coffee Shop   
2         Downtown Toronto               0           Coffee Shop   
4         Downtown Toronto               0                Bakery   
5         Downtown Toronto               0           Coffee Shop   
6         Downtown Toronto               0         Grocery Store   
7         Downtown Toronto               0           Coffee Shop   
8             West Toronto               0              Pharmacy   
10        Downtown Toronto               0           Coffee Shop   
11            West Toronto               0                   Bar   
12            East Toronto               0      Greek Restaurant   
13        Downtown Toronto               0           Coffee Shop   
14            West Toronto               0                  Café   
15            East Toronto               0                   Gym   
16        Downtown Toronto               0           Coffee Shop   
17            East Toronto               0           Coffee Shop   
18         Central Toronto               0                  Park   
20         Central Toronto               0      Department Store   
21         Central Toronto               0                 Trail   
22            West Toronto               0    Mexican Restaurant   
23         Central Toronto               0           Coffee Shop   
24         Central Toronto               0                  Café   
25            West Toronto               0             Gift Shop   
26         Central Toronto               0          Dessert Shop   
27        Downtown Toronto               0                  Café   
28            West Toronto               0                  Café   
30        Downtown Toronto               0                  Café   
31         Central Toronto               0           Coffee Shop   
32        Downtown Toronto               0        Airport Lounge   
34  Downtown Toronto Stn A               0           Coffee Shop   
35        Downtown Toronto               0           Coffee Shop   
36        Downtown Toronto               0           Coffee Shop   
37        Downtown Toronto               0           Coffee Shop   
38   East Toronto Business               0           Yoga Studio   

   2nd Most Common Venue  3rd Most Common Venue  \
0                   Park                    Pub   
1         Clothing Store                  Hotel   
2                   Café             Restaurant   
4            Coffee Shop           Cocktail Bar   
5         Sandwich Place                   Café   
6                   Café                   Park   
7         Clothing Store                   Café   
8                 Bakery            Music Venue   
10              Aquarium                  Hotel   
11                 Diner       Asian Restaurant   
12           Coffee Shop     Italian Restaurant   
13                 Hotel                   Café   
14        Breakfast Spot            Coffee Shop   
15      Sushi Restaurant             Board Shop   
16            Restaurant                   Café   
17                Bakery                   Café   
18           Swim School                   Lake   
20     Food & Drink Shop                   Park   
21         Jewelry Store               Bus Line   
22   Arts & Crafts Store                   Café   
23        Clothing Store                    Spa   
24        Sandwich Place            Coffee Shop   
25        Breakfast Spot       Cuban Restaurant   
26        Sandwich Place            Pizza Place   
27        Sandwich Place                    Bar   
28                   Pub            Pizza Place   
30                   Bar  Vietnamese Restaurant   
31           Pizza Place           Liquor Store   
32      Airport Terminal          Boat or Ferry   
34    Seafood Restaurant     Italian Restaurant   
35                  Café             Restaurant   
36    

__Cluster 2__

In [141]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29          Tennis Court           Summer Camp          Dessert Shop   
33                 Trail            Playground      Department Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
29  Ethiopian Restaurant           Escape Room            Electronics Store   
33           Escape Room     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
33          Dumpling Restaurant            Donut Shop       Doner Restaurant

__Cluster 3__

In [142]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3  East Toronto               2                   Pub                 Trail   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3     Health Food Store              Wine Bar      Doner Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3                 Diner        Discount Store   Distribution Center   

  9th Most Common Venue 10th Most Common Venue  
3               Dog Run    Dumpling Restaurant

__Cluster 4__

In [143]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
9  East York/East Toronto               3                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9     Convenience Store          Intersection   Rental Car Location   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9              Wine Bar    Dim Sum Restaurant           Escape Room   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
9     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

__Cluster 5__

In [144]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               4          Home Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden              Wine Bar          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19  Ethiopian Restaurant           Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

__This is the end of the notebook__. In this notebook some of the knowledge required for a data scientist and some of the skills used by data scientists on a daily basis were shown and applied. The learning material was provided by IBM, and the development of the code and notebook, as well as some notes and editions were carried out by me, [Saulo Villaseñor](https://www.linkedin.com/in/saulo-villase%C3%B1or-60669610a), so that this notebook is available as a reference for anyone who wishes to learn new skills.

## Note: Folium maps are not rendered in GitHub, so please to see the full content displayed with interactive maps click [here](https://nbviewer.jupyter.org/github/saulovillasenor/ibm_data_science_professional_certificate/blob/main/course10_applied_data_science_capstone/week3_neighborhood_segmentation_and_clustering/neighborhoods_toronto.ipynb).